servant-js/src/Servant/JS/Internal.hs

```diff
@@ -87,7 +88,34 @@ defCommonGeneratorOptions = CommonGeneratorOptions
 -- Couldn't work out how to handle zero-width characters.
 --
 -- @TODO: specify better default function name, or throw error?
-toValidFunctionName :: String -> String
+toValidFunctionName :: Text -> Text
+-- @TODO: Cons text
+--
+
+toValidFunctionName t =
+  case T.uncons t of
+    Just (x,xs) ->
+      setFirstChar x `T.cons` T.filter remainder xs 
+    Nothing -> "_"
+  where
+    setFirstChar c = if firstChar c then c else '_'
+    firstChar c = prefixOK c || any (Set.member c) firstLetterOK
+    remainder c = prefixOK c || any (Set.member c) remainderOK
+    -- Valid prefixes
+    prefixOK c = c `elem` ['$','_']
+    -- Unicode character sets
+    firstLetterOK = [ Set.lowercaseLetter
```

Comment:
(I know this wasn't introduced in your change but) I find it a little more intuitive to do a union of sets here and below (and probably more efficient), and then get rid of the `any`s above